couchdb_cluster_admin/suggest_shard_allocation.py

```diff
@@ -1,25 +1,152 @@
 from describe import print_shard_table
 from file_plan import read_plan_file
 from doc_models import ShardAllocationDoc, AllocationSpec
 
-_NodeAllocation = namedtuple('_NodeAllocation', 'i size shards')
 
+class _NodeAllocation(object):
+    def __init__(self, i, size, shards):
+        self.i = i
+        self.size = size
+        self.shards = shards
+
+    def as_tuple(self):
+        return self.i, self.size, self.shards
+
+    def __eq__(self, other):
+        return self.as_tuple() == other.as_tuple()
+
+    def __repr__(self):
+        return '_NodeAllocation({self.i!r}, {self.size!r}, {self.shards!r})'.format(self=self)
+
+
+def suggest_shard_allocation(shard_sizes, n_nodes, n_copies, existing_allocation=None):
+    return Allocator(shard_sizes, n_nodes, n_copies, existing_allocation).suggest_shard_allocation()
+
+
+class Allocator(object):
+    def __init__(self, shard_sizes, n_nodes, n_copies, existing_allocation=None):
+        self.shard_sizes = shard_sizes
+        self.n_nodes = n_nodes
+        self.n_copies = n_copies
+        self.existing_allocation = existing_allocation or ([set()] * self.n_nodes)
+        self.nodes = [_NodeAllocation(i, 0, []) for i in range(self.n_nodes)]
+        self._average_size = sum([size for size, _ in shard_sizes]) * n_copies * 1.0 / n_nodes
+        self._copies_still_in_original_location_by_shard = defaultdict(int)
+        for shards in self.existing_allocation:
+            for shard in shards:
+                self._copies_still_in_original_location_by_shard[shard] += 1
+
+    def suggest_shard_allocation(self):
+        # First distribute, preferring shards' current locations
+        for shard in self._get_shard_sizes_largest_to_smallest():
+            for node in self._select_shard_locations(shard):
+                self._add_shard_to_node(node, shard)
+
+        # Then rebalance
+        self._rebalance_nodes()
+        return self.nodes
+
+    def _get_shard_sizes_largest_to_smallest(self):
+        return [shard for _, shard in reversed(sorted(self.shard_sizes))]
+
+    def _select_shard_locations(self, shard):
+        """
+        Selects best location for n_copies of a given shard, based the allocation so far
+        preferring a shard's existing locations
+
+        returns a list of nodes (_NodeAllocation) that has length n_copies
+        """
+        return sorted(
+            self.nodes,
+            key=lambda node: (shard not in self.existing_allocation[node.i], node.size)
+        )[:self.n_copies]
+
+    @memoized_property
+    def _sizes_by_shard(self):
+        return {shard: size for size, shard in self.shard_sizes}
+
+    def _add_shard_to_node(self, node, shard):
+        node.shards.append(shard)
+        node.size += self._sizes_by_shard[shard]
+
+    def _rebalance_nodes(self):
+        larger_nodes, smaller_nodes = self._split_nodes_by_under_allocated()
+        if not smaller_nodes:
+            return
+        print larger_nodes, smaller_nodes
```

Comment:
is this print intentional? consider making it a log statement with more context?

Reply:
Oops, thanks. Removed